src/Migrations/HSLMigration.php

```diff
@@ -0,0 +1,646 @@
+        'argument_order' => vec[1, 0],
+      ),
+      'substr_replace' => shape(
+        'ns' => HSL_NAMESPACE::Str,
+        'name' => 'splice',
+        'has_overrides' => true,
+      ),
+      'substr' => shape(
+        'ns' => HSL_NAMESPACE::Str,
+        'name' => 'slice',
+        'has_overrides' => true,
+      ),
+      'str_repeat' => shape('ns' => HSL_NAMESPACE::Str, 'name' => 'repeat'),
+      'trim' => shape('ns' => HSL_NAMESPACE::Str, 'name' => 'trim'),
+      'ltrim' => shape('ns' => HSL_NAMESPACE::Str, 'name' => 'trim_left'),
+      'rtrim' => shape('ns' => HSL_NAMESPACE::Str, 'name' => 'trim_right'),
+      'strlen' => shape('ns' => HSL_NAMESPACE::Str, 'name' => 'length'),
+      'sprintf' => shape('ns' => HSL_NAMESPACE::Str, 'name' => 'format'),
+      'str_split' => shape('ns' => HSL_NAMESPACE::Str, 'name' => 'chunk'),
+      'strcmp' => shape('ns' => HSL_NAMESPACE::Str, 'name' => 'compare'),
+      'strcasecmp' => shape('ns' => HSL_NAMESPACE::Str, 'name' => 'compare_ci'),
+      'number_format' =>
+        shape('ns' => HSL_NAMESPACE::Str, 'name' => 'format_number'),
+
+      // Math functions
+      'round' => shape(
+        'ns' => HSL_NAMESPACE::Math,
+        'name' => 'round',
+        'has_overrides' => true,
+      ),
+      'ceil' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'ceil'),
+      'floor' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'floor'),
+      'array_sum' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'sum'),
+      'intdiv' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'int_div'),
+      'exp' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'exp'),
+      'abs' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'abs'),
+      'base_convert' =>
+        shape('ns' => HSL_NAMESPACE::Math, 'name' => 'base_convert'),
+      'pow' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'pow'),
+      'cos' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'cos'),
+      'sin' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'sin'),
+      'tan' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'tan'),
+      'sqrt' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'sqrt'),
+      'log' => shape('ns' => HSL_NAMESPACE::Math, 'name' => 'log'),
+      'min' => shape(
+        'ns' => HSL_NAMESPACE::Math,
+        'name' => 'min',
+        'has_overrides' => true,
+      ),
+      'max' => shape(
+        'ns' => HSL_NAMESPACE::Math,
+        'name' => 'max',
+        'has_overrides' => true,
+      ),
+
+      // Container functions
+      'count' => shape(
+        'ns' => HSL_NAMESPACE::C,
+        'name' => 'count',
+        'argument_order' => vec[0],
+      ),
+    ];
+
+
+  <<__Override>>
+  public function migrateFile(string $path, EditableNode $root): EditableNode {
+    // find all the function calls
+    $nodes = $root->getDescendantsOfType(FunctionCallExpression::class);
+
+    // keep track of any HSL namespaces we may have to add to the top of the file
+    $found_namespaces = dict[];
```

Comment:
Why a dict rather than a `keyset` ?

Reply:
Good point, fixed! I remembered keysets could only contain strings, and didn't think about the fact that the enum was ok to use in a keyset. This simplified a lot of logic.